hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -1269,512 +1315,350 @@ object PruneDeadFields {
     ir: IR,
     requestedType: Type,
     memo: ComputeMutableState,
-  ): BindingEnv[BoxedArrayBuilder[Type]] = {
+    env: BindingEnv[TypeState] = BindingEnv.empty.createAgg.createScan,
+  ): Unit = {
+    def recurMax(ir: IR, childIdx: Int): IndexedSeq[TypeState] =
+      recur(ir, childIdx, ir.getChild(childIdx).asInstanceOf[IR].typ)
+
+    def recurMin(ir: IR, childIdx: Int): IndexedSeq[TypeState] =
+      recur(ir, childIdx, minimal(ir.getChild(childIdx).asInstanceOf[IR].typ))
+
+    def recur(ir: IR, childIdx: Int, requestedType: Type): IndexedSeq[TypeState] =
+      memoizeValueIR(ctx, ir, childIdx, requestedType, memo, env)
+
+    def recurMapMax(ir: IR, childIdx: Int, bindingsMap: mutable.Map[String, TypeState]): Unit =
+      recurMap(ir, childIdx, ir.getChild(childIdx).asInstanceOf[IR].typ, bindingsMap)
+
+    def recurMap(
+      ir: IR,
+      childIdx: Int,
+      requestedType: Type,
+      bindingsMap: mutable.Map[String, TypeState],
+    ): Unit = {
```

Comment:
It feels strange that these return unit while `memoizeValueIR` does not. Was that intentional?

Reply:
The overload of `memoizeValueIR` being called here does return unit. But I agree that's confusing, I'll try to come up with some better names.